projects/concurrency/main_test.go

```diff
@@ -0,0 +1,184 @@
+package main
+
+import (
+	"fmt"
+	"sync"
+	"testing"
+)
+
+func TestCache(t *testing.T) {
+	t.Run("try to get a key from an empty cache", func(t *testing.T) {
+		cache := NewCache[string, int](3)
+
+		_, ok := cache.Get("a")
+
+		if ok != false {
+			t.Errorf("found a key in the cache where there should have been none")
+		}
+	})
+
+	t.Run("add a single key/value to an empty cache", func(t *testing.T) {
+		cache := NewCache[string, int](3)
+
+		insertKey := "a"
+		insertValue := 100
+
+		ok := cache.Put(insertKey, insertValue)
+
+		if ok != false {
+			t.Errorf("found an existing value on the key in the cache, where there should have been no existing value")
+		}
+
+		cacheListElementPointer, ok := cache.entries[insertKey]
+
+		if !ok {
+			t.Errorf("key %v should now exist in the cache but does not", insertKey)
+		}
+
+		// (!) NEED TO LEARN >>> TYPE ASSERTION/CONVERSION
+		// "It performs a type assertion on cacheListElementPointer.Value, checking if the value stored there is a pointer to a CacheValue[string, int] struct."
+		// "If the type assertion is successful, it retrieves the concrete value (the pointer to the CacheValue struct) and assigns it to the cacheValue variable."
+		cacheValue := cacheListElementPointer.Value.(*CacheValue[string, int])
+
+		if cacheValue.value != insertValue {
+			t.Errorf("cacheValue.value: got %v want %v", cacheValue.value, insertValue)
+		}
+	})
+
+	t.Run("add and update a single key/value to an empty cache", func(t *testing.T) {
+		cache := NewCache[string, int](3)
+
+		insertKey := "a"
+		insertValue := 100
+
+		ok := cache.Put(insertKey, insertValue)
+
+		if ok != false {
+			t.Errorf("found an existing value on the key in the cache, where there should have been no existing value")
+		}
+
+		updatedValue := 200
+
+		ok = cache.Put(insertKey, updatedValue)
+
+		if ok != true {
+			t.Errorf("Put should return true to denote a successful updating of that key's value")
+		}
+
+		valuePointer, ok := cache.Get(insertKey)
+
+		if ok != true {
+			t.Errorf("could not find key %v when there should be one", insertKey)
+		}
+
+		// (!) deference the pointer to a value
+		if *valuePointer != updatedValue {
+			t.Errorf("got %v | want %v", *valuePointer, updatedValue)
+		}
+	})
+
+	t.Run("add 3 keys to an empty cache", func(t *testing.T) {
```

Comment:
I'd probably call the test "fill an empty cache" rather than hard-coding 3 - 3 isn't really a magic number in terms of the _interface_ of the cache that you're testing, it's an implementation detail about what "full" means for the cache you happen to be making.